SSA/Projects/InstCombine/ForLean.lean

```diff
@@ -1,6 +1,37 @@
 import Mathlib.Data.Nat.Size -- TODO: remove and get rid of shiftLeft_eq_mul_pow use
 import SSA.Projects.InstCombine.Tactic -- TODO: remove and get rid of ring_nf use
 
+def two_pow_eq_pow_pred_times_two {h : 0 < w} : 2 ^ w = 2 ^ (w-1) * 2 := by
+  simp only [← pow_succ, gt_iff_lt, Nat.ofNat_pos, ne_eq, OfNat.ofNat_ne_one,
+    not_false_eq_true, pow_right_inj]
+  rw [Nat.sub_add_cancel]
+  omega
+
+def two_pow_eq_two_pow_pred_add_two_pow_pred {h : 0 < w} :
+    (2) ^ w = (2^(w-1)) + (2^(w-1)) := by
+  rw [two_pow_eq_pow_pred_times_two, mul_two]
+  omega
+
+def two_pow_gt_two_pow_pred {h : 0 < w} : 2 ^ w > 2 ^ (w - 1) := by
+  simp [@two_pow_eq_two_pow_pred_add_two_pow_pred w h]
+
+def two_pow_pred_lt_two_pow {h : 0 < w} : 2 ^ (w - 1) < 2 ^ w := by
+  simp [two_pow_eq_pow_pred_times_two, h]
+
+@[simp]
+def two_pow_sub_two_pow_pred_eq_two_pow_pred {h : 0 < w} :
+    2 ^ w - 2 ^ (w - 1) = 2 ^ (w - 1) := by
+  simp [@two_pow_eq_two_pow_pred_add_two_pow_pred w h]
+
+def Nat.eq_one_of_mod_two_ne_zero (n : Nat) (hn : n % 2 != 0) : n % 2 = 1 := by
+  simp at hn
+  assumption
+
+def Nat.sub_mod_of_lt (n x : Nat) (hxgt0 : x > 0) (hxltn : x < n) : (n - x) % n = n - x := by
+  rcases n with zero | n <;> simp
```

Comment:
```suggestion
  rcases n with rfl | n <;> simp
```

Use `rfl` when matching with `rcases` to make it clear that you're asking for substitution.